### PR TITLE
Toolbar toggler API

### LINKED ISSUE #1421: How to open block's setting programmatically?
Is there any way to open the block setting programmatically? (Same as clicking 'click to tune' button)

I want to toggle own image block's setting when user clicks on the image.

### PR DESCRIPTION
**Pull request by robonetphy**
resolve #1421 and an updated version of #1442

### CODE REVIEW

File: src/components/modules/api/toolbar.ts [hunk 1]
```diff
@@ -31,4 +32,36 @@ export default class ToolbarAPI extends Module {
   public close(): void {
     this.Editor.Toolbar.close();
   }
+
+  /**
+   * Toggles Toolbar of Current Focused Elements
+   *
+   * @param {boolean} openingState —  opening state of Toolbar
+   */
+  public toggleToolBar(openingState?: boolean): void {
```
Comment (neSpecc) [inline src/components/modules/api/toolbar.ts:41]: it's toggling the Block Settings primarily

File: docs/CHANGELOG.md [hunk 2]
```diff
@@ -1,4 +1,6 @@
 # Changelog
+### 2.19.2
+- `New` -  Toggle API for Block Setting added [#1442](https://github.com/codex-team/editor.js/issues/1421).
```
Comment (neSpecc) [inline docs/CHANGELOG.md:3]:
```suggestion

- `New` -  `toolbar.toggleBlockSettings()` API method added [#1442](https://github.com/codex-team/editor.js/issues/1421). 
```

File: src/components/modules/api/toolbar.ts [hunk 3]
```diff
@@ -15,6 +15,7 @@ export default class ToolbarAPI extends Module {
     return {
       close: (): void => this.close(),
       open: (): void => this.open(),
+      toggleBlockSetting: (openingState?: boolean): void => this.toggleBlockSetting(openingState),
```
Comment (neSpecc) [inline src/components/modules/api/toolbar.ts:18]: toggleBlockSettings

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: src/components/modules/api/toolbar.ts (hunk 3 above)
Comment (robonetphy) [reply src/components/modules/api/toolbar.ts]: solved

File: types/api/toolbar.d.ts [hunk 4]
```diff
@@ -11,4 +11,10 @@ export interface Toolbar {
    * Opens Toolbar
    */
   open(): void;
+
+  /**
+   * Toggles Block Setting of Current Focused Elements
```
Comment (neSpecc) [inline types/api/toolbar.d.ts:16]:
```suggestion
   * Toggles Block Setting of the current block
```

File: src/components/modules/api/toolbar.ts [hunk 5]
```diff
@@ -31,4 +32,36 @@ export default class ToolbarAPI extends Module {
   public close(): void {
     this.Editor.Toolbar.close();
   }
+
+  /**
+   * Toggles Block Setting of Current Focused Elements
```
Comment (neSpecc) [inline src/components/modules/api/toolbar.ts:37]:
```suggestion
   * Toggles Block Setting of the current block
```

File: docs/CHANGELOG.md [hunk 6]
```diff
@@ -1,4 +1,7 @@
 # Changelog
+### 2.19.2
```
Comment (neSpecc) [inline docs/CHANGELOG.md:2]:
```suggestion

### 2.19.2
```

**Review by neSpecc — COMMENTED** [review]
almost ready

**Review by neSpecc — APPROVED** [review]

File: src/components/modules/api/toolbar.ts [hunk 7]
```diff
@@ -31,4 +32,36 @@ export default class ToolbarAPI extends Module {
...
     this.Editor.Toolbar.close();
   }
+
+  /**
+   * Toggles Block Setting of the current block
+   *
+   * @param {boolean} openingState —  opening state of Block Setting
+   */
+  public toggleBlockSettings(openingState?: boolean): void {
+    if (!this.Editor.BlockManager.currentBlockIndex) {
```
Comment (hata6502) [inline src/components/modules/api/toolbar.ts:42]:
```suggestion
    if (this.Editor.BlockManager.currentBlockIndex === -1) {
```

`BlockManager.currentBlockIndex` seems having `-1` or the index of current block.

File: src/components/modules/api/toolbar.ts [hunk 8]
```diff
@@ -31,4 +32,36 @@ export default class ToolbarAPI extends Module {
...
+      _.logLabeled('Could\'t toggle the Toolbar because there is no block selected ', 'warn');
+
+      return;
+    }
+
+    if (!this.Editor.Toolbar.opened) {
```
Comment (hata6502) [inline src/components/modules/api/toolbar.ts:48]: I think this if statement should be moved into: 

```
if (canOpenBlockSettings) {
  this.Editor.BlockSettings.open();
}
```

File: src/components/modules/api/toolbar.ts [hunk 9]
```diff
@@ -31,4 +32,36 @@ export default class ToolbarAPI extends Module {
...
+      this.Editor.Toolbar.open(true, false);
+      this.Editor.Toolbar.plusButton.hide();
+    }
+
+    /** Check that opening state is set or not */
+    const canOpenBlockSettings = (openingState !== undefined) ? openingState : !this.Editor.BlockSettings.opened;
```
Comment (hata6502) [inline src/components/modules/api/toolbar.ts:54]:
```suggestion
    const canOpenBlockSettings = openingState ?? !this.Editor.BlockSettings.opened;
```

File: src/components/modules/api/toolbar.ts [hunk 10]
```diff
@@ -31,4 +32,36 @@ export default class ToolbarAPI extends Module {
...
+
+    /** Check if state same as current state */
+    if (openingState !== undefined && openingState === this.Editor.BlockSettings.opened) {
```
Comment (hata6502) [inline src/components/modules/api/toolbar.ts:57]: `openingState !== undefined` is unused. 

```suggestion
    if (canOpenBlockSettings/* or openingState */ === this.Editor.BlockSettings.opened) {
```

File: src/components/modules/toolbar/blockSettings.ts [hunk 11]
```diff
@@ -170,6 +170,10 @@ export default class BlockSettings extends Module<BlockSettingsNodes> {
     }
 
     this.selection.clearSaved();
+    /**
+     * Remove highlighted content of a Block we are working with
+     */
+    this.Editor.BlockManager.currentBlock.selected = false;
```
Comment (hata6502) [inline src/components/modules/toolbar/blockSettings.ts:176]: When clicking the outside of Editor.js, the following errors are occured: 
```
Uncaught TypeError: Cannot set property 'selected' of undefined
```

```suggestion
    if (this.Editor.BlockManager.currentBlock) {
      this.Editor.BlockManager.currentBlock.selected = false;
    }
```

**Review by hata6502 — CHANGES_REQUESTED** [review]

File: src/components/modules/api/toolbar.ts (hunk 7 above)
Comment (robonetphy) [reply src/components/modules/api/toolbar.ts:42]: Done

File: src/components/modules/api/toolbar.ts (hunk 8 above)
Comment (robonetphy) [reply src/components/modules/api/toolbar.ts:48]: Done

File: src/components/modules/api/toolbar.ts (hunk 9 above)
Comment (robonetphy) [reply src/components/modules/api/toolbar.ts:54]: Done

File: src/components/modules/api/toolbar.ts (hunk 10 above)
Comment (robonetphy) [reply src/components/modules/api/toolbar.ts:57]: Done

File: src/components/modules/toolbar/blockSettings.ts (hunk 11 above)
Comment (robonetphy) [reply src/components/modules/toolbar/blockSettings.ts:176]: Done

**Review by hata6502 — APPROVED** [review]

**Review by neSpecc — APPROVED** [review]

**Review by khaydarov — APPROVED** [review]

**Pull request merged**